### ISSUE TITLE
Have etcd support pre-connected stdio socket pair, pipes and/or unix domain sockets

### ISSUE THREAD

**Issue by z0rb1n0** [coreos/etcd#7296]
Originally requested as from https://groups.google.com/forum/#!topic/etcd-dev/2kQ8PQpANCQ
See Anthony's reply in the thread (his suggestion doesn't seem to work on the version I'm running)

--------------------------------

Sorry if this has been mentioned before but couldn't find anything about it.

In a couple of my projects I'm planning to run etcd as a forked sub-process in a service's process tree:
its life cycle would be entirely tied to the parent process' (it gets signalled, its streams captured and so on), and etcd would take care of global locking/promotions and the like.

It is quite inconvenient (and somewhat inefficient) that the only supported addressing/binding scheme for the client listener is http[s]://bind-address:port (which implies AF_INET[6] sockets), as in our intended setup it adds the need for more security configuration/options in the consumer code base only because we need to talk to etcd through what essentially is a side channel (they're already intimately connected by virtue of inherited environment/file descriptors/known PIDs)

Options I'm suggesting:

- add an "helper" mode, (examples: snmpd's pass_persist , squid's external_acl_type): the parent opens a socketpair/pipe pair, forks and exec etcd with --helper ?; etcd then skips all the listen()/accept()/fork()/clone() cycle and processes http requests on the socket much like it would in an never-ending KeepAlive session.

-- pass etcd a pre-existing file descriptor the aforementioned process can operate on. Something like --clientfd or --clientfd-in and --clientfd-out if you want to separate them

-- simply add support for AF_UNIX binding of the client listener. On Linux in particular this would be an excellent low-hanging fruit as anonymous unix sockets are handled transparently

Is any of this realistically feasible or am I seeing things?

TIA and best regards

F

**Comment by heyitsanthony** [comment]
> his suggestion doesn't seem to work on the version I'm running

What do you mean by not working and what version? `unix://` should be available in >=3.0

**Comment by xiang90** [comment]
@z0rb1n0 Have you tried to specify unix as the protocol as @heyitsanthony suggested? Can you give us an example of the not working case?

**Comment by z0rb1n0** [comment]
Hi.

I ran a couple tests and it turned out that my attempt did something unsupported in both etcd 2.x and 3.x: it looks like the URL parser - correctly - gives up altogether when one tries to pass an absolute socket path, as in unix:///foobar:1234.
I got the same error back from 2.0 and 3.0 I wrongly assumed the feature was not there in 3.1 either. Apologies for the confusion

In 3.0, unix://foobar:1234 obviously works like a charm, although I find it confusing/inconvenient that the path component of the URL is assumed to be relative (qualified URI paths are absolute by definition). I presume that changing that behaviour would break compatibility with a lot of installations tho, so not sure if it's worth looking at it at this point.

Unfortunately it is not possible to use the abstract socket name space on Linux (as the required null char terminates the command line), but that's no big deal either.

All in all, we can achieve most of what is planned so I'm closing this issue as PEBKAC.

Thank you very much and keep up the brilliant work
